README.md
# alcohol/iso3166

A PHP library providing ISO 3166-1 data.

[![Build Status](https://img.shields.io/travis/alcohol/iso3166/master.svg?style=flat-square)](https://travis-ci.org/alcohol/iso3166)
[![License](https://img.shields.io/packagist/l/alcohol/iso3166.svg?style=flat-square)](https://packagist.org/packages/alcohol/iso3166)

## What is ISO 3166-1

> ISO 3166-1 is part of the ISO 3166 standard published by the International Organization for Standardization (ISO), and defines codes for the names of countries, dependent territories, and special areas of geographical interest. The official name of the standard is Codes for the representation of names of countries and their subdivisions – Part 1: Country codes. It defines three sets of country codes:
> * ISO 3166-1 alpha-2 – two-letter country codes which are the most widely used of the three, and used most prominently for the Internet's country code top-level domains (with a few exceptions).
> * ISO 3166-1 alpha-3 – three-letter country codes which allow a better visual association between the codes and the country names than the alpha-2 codes.
> * ISO 3166-1 numeric – three-digit country codes which are identical to those developed and maintained by the United Nations Statistics Division, with the advantage of script (writing system) independence, and hence useful for people or systems using non-Latin scripts.
>
> *-- [Wikipedia](http://en.wikipedia.org/wiki/ISO_3166-1)*

## Installing

``` sh
$ composer require alcohol/iso3166
```

## Using

Code:

``` php
<?php

use Alcohol\ISO3166\ISO3166;

$iso3166 = new ISO3166();

// methods defined by interface Alcohol\ISO3166\DataProvider
$iso3166->getByAlpha2('NL');
$iso3166->getByAlpha3('NLD');
$iso3166->getByNumeric('528');

// methods provided for convenience
$iso3166->getAll();
foreach ($iso3166 as $key => $value) {
    // simple iterator implementation using a generator (uses alpha2 for keys)
}
foreach ($iso3166->listBy(ISO3166::KEY_ALPHA3) as $key => $value) {
    // use a specific generator to iterate using other available keys
    // constants available are KEY_ALPHA2, KEY_ALPHA3, KEY_NUMERIC
}
```

The data returned is in the format of:

```
Array
(
    [name] => Netherlands
    [alpha2] => NL
    [alpha3] => NLD
    [numeric] => 528
    [currency] => EUR
)
```

## Contributing

Feel free to submit a pull request or create an issue.

## License

alcohol/iso3166 is licensed under the MIT license.

## Source(s)

* [ISO 3166-1](http://en.wikipedia.org/wiki/ISO_3166-1) by [Wikipedia](http://www.wikipedia.org) licensed under [CC BY-SA 3.0 Unported License](http://en.wikipedia.org/wiki/Wikipedia:Text_of_Creative_Commons_Attribution-ShareAlike_3.0_Unported_License)
* [www.iso.org](http://www.iso.org)


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file.
This project adheres to [Semantic Versioning](http://semver.org/).

## [4.0.0] - 2015-12-11

* Changed: added a CHANGELOG.
* Changed: dropped support for PHP 5.4.
* Changed: can now iterate the data instance.
* Changed: added a `DataProvider` interface.

## [3.0.0] - 2015-09-04

* Changed: exceptions better reflect the error they represent.
* Changed: reviewed all names vs scraped content from www.iso.org.

## [2.2.0] - 2015-07-31

* Changed: dropped support for PHP 5.3, switched to short array notation.
* Changed: updated Lithuanian currency to EUR.

[4.0.0]: https://github.com/alcohol/iso3166/compare/3.0.0...4.0.0
[3.0.0]: https://github.com/alcohol/iso3166/compare/2.2.0...3.0.0
[2.2.0]: https://github.com/alcohol/iso3166/compare/2.1.3...2.2.0


tests/ISO3166Test.php
<?php

/**
 * (c) Rob Bast <rob.bast@gmail.com>
 *
 * For the full copyright and license information, please view
 * the LICENSE file that was distributed with this source code.
 */

namespace Alcohol\Tests;

use Alcohol\ISO3166;

class ISO3166Test extends \PHPUnit_Framework_TestCase
{
    /**
     * @testdox Calling getByAlpha2 with an invalid alpha2 throws a InvalidArgumentException.
     * @dataProvider invalidAlpha2Provider
     * @param string $alpha2
     * @expectedException \InvalidArgumentException
     * @expectedExceptionMessageRegExp /^Not a valid alpha2: .*$/
     */
    public function testGetByAlpha2Invalid($alpha2)
    {
        $iso3166 = new ISO3166;
        $iso3166->getByAlpha2($alpha2);
    }

    /**
     * @testdox Calling getByAlpha2 with an unknown alpha2 throws a RuntimeException.
     * @expectedException \RuntimeException
     * @expectedExceptionMessage ISO 3166-1 does not contain: ZZ
     */
    public function testGetByAlpha2Unknown()
    {
        $iso3166 = new ISO3166;
        $iso3166->getByAlpha2('ZZ');
    }

    /**
     * @testdox Calling getByAlpha2 with a known alpha2 returns an associative array with the data.
     * @dataProvider alpha2Provider
     * @param string $alpha2
     * @param array $expected
     */
    public function testGetByAlpha2($alpha2, array $expected)
    {
        $iso3166 = new ISO3166;
        $this->assertEquals($expected, $iso3166->getByAlpha2($alpha2));
    }

    /**
     * @testdox Calling getByAlpha3 with an invalid alpha3 throws a InvalidArgumentException.
     * @param string $alpha3
     * @dataProvider invalidAlpha3Provider
     * @expectedException \InvalidArgumentException
     * @expectedExceptionMessageRegExp /^Not a valid alpha3: .*$/
     */
    public function testGetByAlpha3Invalid($alpha3)
    {
        $iso3166 = new ISO3166;
        $iso3166->getByAlpha3($alpha3);
    }

    /**
     * @testdox Calling getByAlpha3 with an unknown alpha3 throws a RuntimeException.
     * @expectedException \RuntimeException
     * @expectedExceptionMessage ISO 3166-1 does not contain: ZZZ
     */
    public function testGetByAlpha3Unknown()
    {
        $iso3166 = new ISO3166;
        $iso3166->getByAlpha3('ZZZ');
    }

    /**
     * @testdox Calling getByAlpha3 with a known alpha3 returns an associative array with the data.
     * @dataProvider alpha3Provider
     * @param string $alpha3
     * @param array $expected
     */
    public function testGetByAlpha3($alpha3, array $expected)
    {
        $iso3166 = new ISO3166;
        $this->assertEquals($expected, $iso3166->getByAlpha3($alpha3));
    }

    /**
     * @testdox Calling getByNumeric with an invalid numeric throws a InvalidArgumentException.
     * @param string $numeric
     * @dataProvider invalidNumericProvider
     * @expectedException \InvalidArgumentException
     * @expectedExceptionMessageRegExp /^Not a valid numeric: .*$/
     */
    public function testGetByNumericInvalid($numeric)
    {
        $iso3166 = new ISO3166;
        $iso3166->getByNumeric($numeric);
    }

    /**
     * @testdox Calling getByNumeric with an unknown numeric throws a RuntimeException.
     * @expectedException \RuntimeException
     * @expectedExceptionMessage ISO 3166-1 does not contain: 000
     */
    public function testGetByNumericUnknown()
    {
        $iso3166 = new ISO3166;
        $iso3166->getByNumeric('000');
    }

    /**
     * @testdox Calling getByNumeric with a known numeric returns an associative array with the data.
     * @dataProvider numericProvider
     * @param string $numeric
     * @param array $expected
     */
    public function testGetByNumeric($numeric, $expected)
    {
        $iso3166 = new ISO3166;
        $this->assertEquals($expected, $iso3166->getByNumeric($numeric));
    }

    /**
     * @testdox Calling getAll returns an array with all elements.
     */
    public function testGetAll()
    {
        $iso3166 = new ISO3166;
        $this->assertInternalType('array', $iso3166->getAll());
        $this->assertCount(249, $iso3166->getAll());
    }

    /**
     * @return array
     */
    public function invalidAlpha2Provider()
    {
        return array(array('Z'), array('ZZZ'), array(1), array(123));
    }

    /**
     * @return array
     */
    public function alpha2Provider()
    {
        $countries = $this->getCountries();

        return array_reduce(
            $countries,
            function (array $carry, array $country) {
                $carry[] = array($country['alpha2'], $country);
                return $carry;
            },
            array()
        );
    }

    /**
     * @return array
     */
    public function invalidAlpha3Provider()
    {
        return array(array('ZZ'), array('ZZZZ'), array(12), array(1234));
    }

    /**
     * @return array
     */
    public function alpha3Provider()
    {
        $countries = $this->getCountries();

        return array_reduce(
            $countries,
            function (array $carry, array $country) {
                $carry[] = array($country['alpha3'], $country);
                return $carry;
            },
            array()
        );
    }

    /**
     * @return array
     */
    public function invalidNumericProvider()
    {
        return array(array('00'), array('0000'), array('ZZ'), array('ZZZZ'));
    }

    /**
     * @return array
     */
    public function numericProvider()
    {
        $countries = $this->getCountries();

        return array_reduce(
            $countries,
            function (array $carry, array $country) {
                $carry[] = array($country['numeric'], $country);
                return $carry;
            },
            array()
        );
    }

    /**
     * @return array
     */
    private function getCountries()
    {
        $reflected = new \ReflectionClass('Alcohol\ISO3166');
        $countries = $reflected->getProperty('countries');
        $countries->setAccessible(true);

        return $countries->getValue(new ISO3166);
    }
}


ISO3166Test.php
<?php

/*
 * (c) Rob Bast <rob.bast@gmail.com>
 *
 * For the full copyright and license information, please view
 * the LICENSE file that was distributed with this source code.
 */

namespace Alcohol\ISO3166;

class ISO3166Test extends \PHPUnit_Framework_TestCase
{
    /**
     * @testdox Calling getByAlpha2 with an invalid alpha2 throws a DomainException.
     * @dataProvider invalidAlpha2Provider
     * @expectedException \DomainException
     * @expectedExceptionMessageRegExp /^Not a valid alpha2: .*$/
     *
     * @param string $alpha2
     */
    public function testGetByAlpha2Invalid($alpha2)
    {
        $iso3166 = new ISO3166();
        $iso3166->getByAlpha2($alpha2);
    }

    /**
     * @testdox Calling getByAlpha2 with an unknown alpha2 throws a OutOfBoundsException.
     * @expectedException \OutOfBoundsException
     * @expectedExceptionMessage ISO 3166-1 does not contain: ZZ
     */
    public function testGetByAlpha2Unknown()
    {
        $iso3166 = new ISO3166();
        $iso3166->getByAlpha2('ZZ');
    }

    /**
     * @testdox Calling getByAlpha2 with a known alpha2 returns an associative array with the data.
     * @dataProvider alpha2Provider
     *
     * @param string $alpha2
     * @param array $expected
     */
    public function testGetByAlpha2($alpha2, array $expected)
    {
        $iso3166 = new ISO3166();
        $this->assertEquals($expected, $iso3166->getByAlpha2($alpha2));
    }

    /**
     * @testdox Calling getByAlpha3 with an invalid alpha3 throws a DomainException.
     * @dataProvider invalidAlpha3Provider
     * @expectedException \DomainException
     * @expectedExceptionMessageRegExp /^Not a valid alpha3: .*$/
     *
     * @param string $alpha3
     */
    public function testGetByAlpha3Invalid($alpha3)
    {
        $iso3166 = new ISO3166();
        $iso3166->getByAlpha3($alpha3);
    }

    /**
     * @testdox Calling getByAlpha3 with an unknown alpha3 throws a OutOfBoundsException.
     * @expectedException \OutOfBoundsException
     * @expectedExceptionMessage ISO 3166-1 does not contain: ZZZ
     */
    public function testGetByAlpha3Unknown()
    {
        $iso3166 = new ISO3166();
        $iso3166->getByAlpha3('ZZZ');
    }

    /**
     * @testdox Calling getByAlpha3 with a known alpha3 returns an associative array with the data.
     * @dataProvider alpha3Provider
     *
     * @param string $alpha3
     * @param array $expected
     */
    public function testGetByAlpha3($alpha3, array $expected)
    {
        $iso3166 = new ISO3166();
        $this->assertEquals($expected, $iso3166->getByAlpha3($alpha3));
    }

    /**
     * @testdox Calling getByNumeric with an invalid numeric throws a DomainException.
     * @dataProvider invalidNumericProvider
     * @expectedException \DomainException
     * @expectedExceptionMessageRegExp /^Not a valid numeric: .*$/
     *
     * @param string $numeric
     */
    public function testGetByNumericInvalid($numeric)
    {
        $iso3166 = new ISO3166();
        $iso3166->getByNumeric($numeric);
    }

    /**
     * @testdox Calling getByNumeric with an unknown numeric throws a OutOfBoundsException.
     * @expectedException \OutOfBoundsException
     * @expectedExceptionMessage ISO 3166-1 does not contain: 000
     */
    public function testGetByNumericUnknown()
    {
        $iso3166 = new ISO3166();
        $iso3166->getByNumeric('000');
    }

    /**
     * @testdox Calling getByNumeric with a known numeric returns an associative array with the data.
     * @dataProvider numericProvider
     *
     * @param string $numeric
     * @param array $expected
     */
    public function testGetByNumeric($numeric, $expected)
    {
        $iso3166 = new ISO3166();
        $this->assertEquals($expected, $iso3166->getByNumeric($numeric));
    }

    /**
     * @testdox Calling getAll returns an array with all elements.
     */
    public function testGetAll()
    {
        $iso3166 = new ISO3166();
        $this->assertInternalType('array', $iso3166->getAll());
        $this->assertCount(249, $iso3166->getAll());
    }

    /**
     * @testdox Iterating the instance should behave as expected.
     */
    public function testIterator()
    {
        try {
            new ISO3166('foo');
        } catch (\Exception $e) {
            $this->assertInstanceOf('DomainException', $e);
            $this->assertRegExp('{Invalid value given for \$iteratorKey, got "\w++", expected one of:(?: \w++,?)+}', $e->getMessage());
        }


        $iso3166 = new ISO3166();
        $i = 0;
        foreach ($iso3166 as $key => $value) {
            ++$i;
        }
        $this->assertEquals(count($iso3166->getAll()), $i, 'Compare iterator count to count(getAll()).');
    }

    /**
     * @testdox Iterator should be count()'able.
     */
    public function testCountable()
    {
        $iso3166 = new ISO3166();
        $this->assertEquals(count($iso3166->getAll()), count($iso3166));
    }

    /**
     * @return array
     */
    public function invalidAlpha2Provider()
    {
        return [['Z'], ['ZZZ'], [1], [123]];
    }

    /**
     * @return array
     */
    public function alpha2Provider()
    {
        return $this->getCountries('alpha2');
    }

    /**
     * @return array
     */
    public function invalidAlpha3Provider()
    {
        return [['ZZ'], ['ZZZZ'], [12], [1234]];
    }

    /**
     * @return array
     */
    public function alpha3Provider()
    {
        return $this->getCountries('alpha3');
    }

    /**
     * @return array
     */
    public function invalidNumericProvider()
    {
        return [['00'], ['0000'], ['ZZ'], ['ZZZZ']];
    }

    /**
     * @return array
     */
    public function numericProvider()
    {
        return $this->getCountries('numeric');
    }

    /**
     * @param string $indexedBy
     *
     * @return array
     */
    private function getCountries($indexedBy)
    {
        $reflected = new \ReflectionClass('Alcohol\ISO3166\ISO3166');
        $countries = $reflected->getProperty('countries');
        $countries->setAccessible(true);
        $countries = $countries->getValue(new ISO3166());

        return array_reduce(
            $countries,
            function (array $carry, array $country) use ($indexedBy) {
                $carry[] = [$country[$indexedBy], $country];

                return $carry;
            },
            []
        );
    }
}
